lib/ecto_adapters_dynamodb/query.ex

```diff
@@ -90,58 +101,85 @@ defmodule Ecto.Adapters.DynamoDB.Query do
         query = construct_search(index, search, opts)
 
         if op == :in do
-          responses_element = "Responses"
-          response_map = %{responses_element => %{table => []}}
+          response_map = %{@responses_element_key => %{table => []}}
 
           Enum.reduce(hash_values, response_map, fn(hash_value, acc) ->
             # When receiving a list of values to query on, construct a custom query for each of those values to pass into do_fetch_recursive/1.
-            %{"Items" => items} = Kernel.put_in(query, [:expression_attribute_values, :hash_key], hash_value)
+            %{"Items" => items} = put_in(query, [:expression_attribute_values, :hash_key], hash_value)
                                   |> (do_fetch_recursive).()
 
-            Kernel.put_in(acc, [responses_element, table], acc[responses_element][table] ++ items)
+            put_in(acc, [@responses_element_key, table], acc[@responses_element_key][table] ++ items)
           end)
         else
           do_fetch_recursive.(query)
         end
       :scan ->
         maybe_scan(table, search, opts)
     end
+  end
+
+  @spec filter(dynamo_response,  non_indexed_filters, String.t) :: dynamo_response | %{}
+  defp filter(%{"Item" => item} = result, non_indexed_filters, _) do
+    if passes_filter?(item, non_indexed_filters),
+      do: result,
+      else: %{}
+  end
+  defp filter(%{@responses_element_key => items} = result, non_indexed_filters, table) do
+    filtered_results = Enum.filter(items[table], &passes_filter?(&1, non_indexed_filters))
+    put_in(result, [@responses_element_key, table], filtered_results)
+  end
+
 
-    results
+  # @spec passes_filter?(dynamo_response, non_indexed_filters) :: boolean()
+  defp passes_filter?(item, [{ logical_op, [ deeper_clauses, filter ] }]) do
+    case logical_op do
+      :and -> evaluate_filter_expression(item, filter) and passes_filter?(item, deeper_clauses)
```

Comment:
This is indeed complex, tree-ed logic. We may want a test with some parenthetical and and or.

Reply:
Can you give an example?

Reply:
Yes, something like `where (A or B) and (B or C)` where the capital letters are boolean filter expressions.

Reply:
To help me refactor the work I had done on this last week, I started off by using some simplified data structures and a little test module with a stripped down version of the recursive function (it just deals with the filter, not with the item-to-be-filtered):

```
defmodule R do
  def run(tree), do: tree |> recurse()

  defp recurse([{ boolean_op, [ deeper_clauses, filter ] }]) do
    case boolean_op do
      :and -> filter and recurse(deeper_clauses)
      :or -> filter or recurse(deeper_clauses)
    end
  end
  defp recurse([{ _, [ filter ] }]), do: filter
  defp recurse(clauses_or_filter), do: recurse([ clauses_or_filter ])
end

tree1 =
  [
    and: [
      {
        :or,
        [
          {
            :and,
            [
              true
            ]
          },
          false
        ]
      },
      true
    ]
  ]

tree2 =
  [
    or: [
      {
        :and,
        [
          {
            :or,
            [
              {
                :and,
                [
                  {
                    :and,
                    [
                      false
                    ]
                  },
                  false
                ]
              },
              true
            ]
          },
          true
        ]
      },
      false
    ]
  ]

tree3 =
  [
    or: [
      false
    ]
  ]

IO.inspect R.run(tree1)
IO.inspect R.run(tree2)
IO.inspect R.run(tree3)
```

I exploded the trees to make them "easier" to read - you could toss this in an .exs file to experiment with it.  I experimented with the changing the operators and the booleans, but these examples (and the variations I played with) all worked as I expected.  Is this along the lines of what you meant?  By all means, please to try to break it.

Reply:
What I meant by test is a test in the testing suite that makes a query of the kind I suggested (with some parentheticals that can generate a deeper structure). Even though the function may be correct, we want to be sure that the data structres we are parsing actually match and the query works.

Reply:
Ok - I'll throw some big fatty tests in.